onnxruntime/core/providers/cuda/nn/conv_transpose.cc

```diff
@@ -44,26 +44,38 @@ Status ConvTranspose<T, NHWC>::ComputeInternal(OpKernelContext* context) const {
 }
 
 template <typename T, bool NHWC>
-Status ConvTranspose<T, NHWC>::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc, bool& is_packed,
-                                       [[maybe_unused]] PrePackedWeights* prepacked_weights) {
+Status ConvTranspose<T, NHWC>::PrePack([[maybe_unused]] const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+                                       bool& is_packed, [[maybe_unused]] PrePackedWeights* prepacked_weights) {
   is_packed = false;
   // only layout of weight input is adjusted via PrePack
-  if (NHWC) {  // InputTensors::IN_W
+  if constexpr (NHWC) {  // InputTensors::IN_W
     if (input_idx == 1) {
-      // Transpose from {M, C/group, kH, kW} to {M, kH, kW, C/group}
       auto orig_shape = tensor.Shape();
+      const auto rank = orig_shape.NumDimensions();
+
+      InlinedVector<size_t> perm;
+      TensorShapeVector new_dims;
+
+      if (rank == 3) {
+        // Transpose from {C, M/group, k1} to {M/group, k1, C}
+        perm = {1, 2, 0};
+        new_dims = TensorShapeVector{orig_shape[1], orig_shape[2], orig_shape[0]};
+      } else if (rank == 4) {
+        // Transpose from {C, M/group, kH, kW} to {M/group, kH, kW, C}
+        perm = {1, 2, 3, 0};
+        new_dims = TensorShapeVector{orig_shape[1], orig_shape[2], orig_shape[3], orig_shape[0]};
+      } else if (rank == 5) {
+        // Transpose from {C, M/group, k1, k2, k3} to {M/group, k1, k2, k3, C}
+        perm = {1, 2, 3, 4, 0};
+        new_dims = TensorShapeVector{orig_shape[1], orig_shape[2], orig_shape[3], orig_shape[4], orig_shape[0]};
+      }
```

Comment:
@gedoensmax do these changes make sense to you? Looked like the Conv PrePack was copied and not adjusted for the differences in the ConvTranspose spec. The unit tests didn't reveal the issue as all the W values were 1 (so how they were transposed didn't change the output), and both the N and C dims were 1 (comparing incorrect dims for equality didn't fail).

Reply:
Seems to be implementation dependent as to what is moved. I _think_ the 'C' and 'M' values _could_ be viewed as 'input channels' and 'output channels' respectively, and doing so allows me to make sense of this. Whether the input or output channels related data in W is moved to be last seems to be implementation dependent. XNNPACK moves the C. CUDA moves the M. hopefully this interpretation is valid.